# Recommendation System

A Recommendation System is a type of information filtering system that suggests relevant items to users based on their preferences, behavior, or past interactions.

It is widely used in platforms such as:
- E-commerce websites
- Streaming services
- Social media platforms
- Online learning platforms

The main goal of a recommendation system is to **predict user preferences and recommend items that the user is likely to be interested in**.

---

## Examples of Recommendation Systems

- Amazon recommending products based on past purchases
- Netflix recommending movies and shows
- Spotify recommending music playlists
- YouTube recommending videos

---

## Types of Recommendation Systems

### 1. Content-Based Filtering

Content-based filtering recommends items based on the **similarity between item features and user preferences**.

Example:
If a user likes **action movies**, the system will recommend other **action movies**.

Characteristics:
- Uses item features
- Based on user history
- No need for other users' data

---

### 2. Collaborative Filtering

Collaborative filtering recommends items based on **similar users or similar user behavior**.

Example:
If **User A and User B** like similar movies, then movies liked by **User B** may be recommended to **User A**.

Types of Collaborative Filtering:

- **User-Based Collaborative Filtering**
- **Item-Based Collaborative Filtering**

---

### 3. Hybrid Recommendation System

Hybrid systems combine **Content-Based Filtering and Collaborative Filtering** to improve recommendation accuracy.

Example:
Netflix uses a **hybrid recommendation system**.

---

## Steps in Building a Recommendation System

1. Data Collection
2. Data Preprocessing
3. Feature Extraction
4. Model Building
5. Recommendation Generation
6. Evaluation

---

## Applications of Recommendation Systems

- E-commerce product recommendations
- Movie and music recommendations
- News article recommendations
- Online advertising
- Job recommendations

---

## Advantages

- Improves user experience
- Helps users discover relevant content
- Increases engagement and sales

---

## Challenges

- Cold Start Problem (new users or items)
- Data sparsity
- Scalability issues



## Plan of Action

To build a Movie Recommendation System, we will follow these steps:

1. **Import Libraries**
   - Import required Python libraries such as pandas, numpy, sklearn, etc.

2. **Load the Datasets**
   - Load the **movies dataset** and **credits dataset**.

3. **Merge the Datasets**
   - Join both datasets using the common column (movie_id or title).

4. **Select Important Columns**
   - Keep only useful columns for the recommendation system.

   Selected columns:
   - movie_id
   - title
   - overview
   - genres
   - keywords
   - cast
   - crew
   - release_date

5. **Data Preprocessing**
   - Handle missing values.
   - Convert text data into proper format.
   - Extract useful information (e.g., director from crew).

6. **Feature Engineering**
   - Combine important columns to create a new feature called **tags**.

7. **Text Vectorization**
   - Convert the text data into numerical vectors using techniques such as **Bag of Words** or **TF-IDF**.

8. **Calculate Similarity**
   - Use **Cosine Similarity** to measure similarity between movies.

9. **Build Recommendation Function**
   - Create a function that takes a movie name and returns similar movies.

10. **Test the Recommendation System**
   - Check if the system recommends relevant movies based on user input.

In [2]:
import pandas as pd
import numpy as np

In [3]:
movies = pd.read_csv("tmdb_5000_movies.csv")
movies.head()

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.sonypictures.com/movies/spectre/,206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...","[{""iso_3166_1"": ""GB"", ""name"": ""United Kingdom""...",2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466
3,250000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 80, ""nam...",http://www.thedarkknightrises.com/,49026,"[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,...",en,The Dark Knight Rises,Following the death of District Attorney Harve...,112.312950,"[{""name"": ""Legendary Pictures"", ""id"": 923}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-07-16,1084939099,165.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,The Legend Ends,The Dark Knight Rises,7.6,9106
4,260000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://movies.disney.com/john-carter,49529,"[{""id"": 818, ""name"": ""based on novel""}, {""id"":...",en,John Carter,"John Carter is a war-weary, former military ca...",43.926995,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}]","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-03-07,284139100,132.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Lost in our world, found in another.",John Carter,6.1,2124


In [4]:
movies.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count'],
      dtype='object')

## Movie Dataset Features

The movie dataset contains several columns that provide information about each movie.  
These features can be used to build a **movie recommendation system**.

### List of Columns

- **budget** – The total budget used to produce the movie.
- **genres** – The categories of the movie such as Action, Comedy, Drama, etc.
- **homepage** – The official website link of the movie.
- **id** – Unique identifier assigned to each movie.
- **keywords** – Important keywords related to the movie’s theme or story.
- **original_language** – The original language in which the movie was produced.
- **original_title** – The original name of the movie.
- **overview** – A short description or summary of the movie's storyline.
- **popularity** – A score representing how popular the movie is.
- **production_companies** – Companies involved in producing the movie.
- **production_countries** – Countries where the movie was produced.
- **release_date** – The date when the movie was released.
- **revenue** – Total revenue generated by the movie.
- **runtime** – Duration of the movie in minutes.
- **spoken_languages** – Languages spoken in the movie.
- **status** – Indicates whether the movie is released, planned, etc.
- **tagline** – A short catchy phrase used for marketing the movie.
- **title** – The main title of the movie.
- **vote_average** – Average rating given by users.
- **vote_count** – Total number of users who rated the movie.

In [5]:
credit = pd.read_csv("tmdb_5000_credits.csv")
credit.head()

,movie_id,title,cast,crew
0,19995,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,206647,Spectre,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,49026,The Dark Knight Rises,"[{""cast_id"": 2, ""character"": ""Bruce Wayne / Ba...","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,49529,John Carter,"[{""cast_id"": 5, ""character"": ""John Carter"", ""c...","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."


In [6]:
movies.describe()

,budget,id,popularity,revenue,runtime,vote_average,vote_count
count,4.803000e+03,4803.000000,4803.000000,4.803000e+03,4801.000000,4803.000000,4803.000000
mean,2.904504e+07,57165.484281,21.492301,8.226064e+07,106.875859,6.092172,690.217989
std,4.072239e+07,88694.614033,31.816650,1.628571e+08,22.611935,1.194612,1234.585891
min,0.000000e+00,5.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000
25%,7.900000e+05,9014.500000,4.668070,0.000000e+00,94.000000,5.600000,54.000000
50%,1.500000e+07,14629.000000,12.921594,1.917000e+07,103.000000,6.200000,235.000000
75%,4.000000e+07,58610.500000,28.313505,9.291719e+07,118.000000,6.800000,737.000000
max,3.800000e+08,459488.000000,875.581305,2.787965e+09,338.000000,10.000000,13752.000000


In [7]:
movies[movies["popularity"]==875.581305]

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
546,74000000,"[{""id"": 10751, ""name"": ""Family""}, {""id"": 16, ""...",http://www.minionsmovie.com/,211672,"[{""id"": 3487, ""name"": ""assistant""}, {""id"": 179...",en,Minions,"Minions Stuart, Kevin and Bob are recruited by...",875.581305,"[{""name"": ""Universal Pictures"", ""id"": 33}, {""n...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2015-06-17,1156730962,91.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Before Gru, they had a history of bad bosses",Minions,6.4,4571


In [8]:
credit.columns

Index(['movie_id', 'title', 'cast', 'crew'], dtype='object')

## Credits Dataset Features

The credits dataset contains information about the **actors and crew members involved in each movie**.  
This data helps in understanding who worked in the movie and can be useful for building a **content-based recommendation system**.

### List of Columns

- **movie_id** – A unique identifier for each movie. It is used to link this dataset with the main movie dataset.

- **title** – The name of the movie.

- **cast** – A list of actors who acted in the movie along with their roles.

- **crew** – A list of crew members involved in the movie such as the **director, producer, writer, cinematographer**, etc.

# merge both dataset

In [9]:
data = movies.merge(credit, on="title")
data.head()

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,...,runtime,spoken_languages,status,tagline,title,vote_average,vote_count,movie_id,cast,crew
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...",...,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800,19995,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...",...,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500,285,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.sonypictures.com/movies/spectre/,206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...",...,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466,206647,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,250000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 80, ""nam...",http://www.thedarkknightrises.com/,49026,"[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,...",en,The Dark Knight Rises,Following the death of District Attorney Harve...,112.312950,"[{""name"": ""Legendary Pictures"", ""id"": 923}, {""...",...,165.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,The Legend Ends,The Dark Knight Rises,7.6,9106,49026,"[{""cast_id"": 2, ""character"": ""Bruce Wayne / Ba...","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,260000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://movies.disney.com/john-carter,49529,"[{""id"": 818, ""name"": ""based on novel""}, {""id"":...",en,John Carter,"John Carter is a war-weary, former military ca...",43.926995,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}]",...,132.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Lost in our world, found in another.",John Carter,6.1,2124,49529,"[{""cast_id"": 5, ""character"": ""John Carter"", ""c...","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."


In [10]:
data["genres"][0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

## Important Columns After Merging Datasets

After merging the movies and credits datasets, we select only the columns that help describe the movie content.

### Selected Columns

- **movie_id** – Unique identifier for each movie.
- **title** – Name of the movie.
- **overview** – Short description of the movie story.
- **genres** – Categories of the movie (Action, Comedy, Drama, etc.).
- **keywords** – Important words describing the movie theme.
- **cast** – Main actors in the movie.
- **crew** – Information about crew members (especially the director).
- **release_date** - The date when the movie was released.

These columns are important because they describe the **story, theme, actors, and direction of the movie**, which helps in building a **content-based recommendation system**.

In [11]:
data = data[["genres","id","keywords","overview","release_date","title","cast","crew"]]

In [12]:
data.shape

(4809, 8)

In [13]:
data.describe()

,id
count,4809.000000
mean,57120.571429
std,88653.369849
min,5.000000
25%,9012.000000
50%,14624.000000
75%,58595.000000
max,459488.000000


In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   genres        4809 non-null   object
 1   id            4809 non-null   int64 
 2   keywords      4809 non-null   object
 3   overview      4806 non-null   object
 4   release_date  4808 non-null   object
 5   title         4809 non-null   object
 6   cast          4809 non-null   object
 7   crew          4809 non-null   object
dtypes: int64(1), object(7)
memory usage: 300.7+ KB


In [15]:
data.isnull().sum()

genres          0
id              0
keywords        0
overview        3
release_date    1
title           0
cast            0
crew            0
dtype: int64

In [16]:
data.dropna(inplace=True)

In [17]:
data.shape

(4805, 8)

In [18]:
data.duplicated().sum()

np.int64(0)

# Preprocessing Steps

## Extracting Useful Information from Genres

The **genres** column in the dataset contains information about the categories of the movie such as Action, Comedy, Drama, etc.

However, the data in this column is stored in a **JSON-like format**, where each genre is represented as a dictionary containing an **id** and **name**.

Example:

[
{'id': 28, 'name': 'Action'},
{'id': 12, 'name': 'Adventure'},
{'id': 14, 'name': 'Fantasy'}
]

For building the recommendation system, we **do not need the id**.  
We only need the **genre names** because they describe the movie category.

### Goal

Extract only the **genre names** from the genres column.

Example Output:

['Action', 'Adventure', 'Fantasy']

These extracted genre names will later be **combined with other features (keywords, cast, overview, etc.)** to create a **tags column** used for calculating movie similarity.

In [19]:
import ast

In [20]:
data["genres"][0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

In [21]:
def genre_extract(col):
    L=[]
    for i in ast.literal_eval(col):
        L.append(i["name"])
    return L

In [22]:
data["genres"]= data["genres"].apply(genre_extract)

## ast.literal_eval()

`ast.literal_eval()` is a function from Python's **ast (Abstract Syntax Tree)** module that safely evaluates a string containing a Python literal or data structure.

It converts a **string representation of a Python object** into the actual Python object.

### Example

```python
import ast

text = "[1, 2, 3]"
result = ast.literal_eval(text)

print(result)
print(type(result))

[1, 2, 3]
<class 'list'>

In [23]:
data.head()

,genres,id,keywords,overview,release_date,title,cast,crew
0,"[Action, Adventure, Fantasy, Science Fiction]",19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...","In the 22nd century, a paraplegic Marine is di...",2009-12-10,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,"[Adventure, Fantasy, Action]",285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...","Captain Barbossa, long believed to be dead, ha...",2007-05-19,Pirates of the Caribbean: At World's End,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,"[Action, Adventure, Crime]",206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",A cryptic message from Bond’s past sends him o...,2015-10-26,Spectre,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,"[Action, Crime, Drama, Thriller]",49026,"[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,...",Following the death of District Attorney Harve...,2012-07-16,The Dark Knight Rises,"[{""cast_id"": 2, ""character"": ""Bruce Wayne / Ba...","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,"[Action, Adventure, Science Fiction]",49529,"[{""id"": 818, ""name"": ""based on novel""}, {""id"":...","John Carter is a war-weary, former military ca...",2012-03-07,John Carter,"[{""cast_id"": 5, ""character"": ""John Carter"", ""c...","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."


## Extracting Keywords

The **keywords** column contains important words related to the movie's theme or story.

The data is stored in a JSON-like format, so we need to extract only the **keyword names** for further processing.

In [24]:
data["keywords"][0]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [25]:
import ast

def extract_keywords(text):
    L = []
    for i in ast.literal_eval(text):
        L.append(i['name'])
    return L

data['keywords'] = data['keywords'].apply(extract_keywords)



## Processing Release Date

The **release_date** column contains the date when the movie was released.

Example format:
2009-12-18

For the recommendation system, we usually extract only the **release year** because movies released in similar time periods may have similar styles or trends.

Example:
2009-12-18 → 2009

In [26]:
data["release_date"].dtypes

dtype('O')

In [27]:
data["release_date"]=pd.to_datetime(data["release_date"])

In [28]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4805 entries, 0 to 4808
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   genres        4805 non-null   object        
 1   id            4805 non-null   int64         
 2   keywords      4805 non-null   object        
 3   overview      4805 non-null   object        
 4   release_date  4805 non-null   datetime64[ns]
 5   title         4805 non-null   object        
 6   cast          4805 non-null   object        
 7   crew          4805 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 466.9+ KB


In [29]:
data["release_date"] = data["release_date"].dt.year

In [30]:
data.head()

,genres,id,keywords,overview,release_date,title,cast,crew
0,"[Action, Adventure, Fantasy, Science Fiction]",19995,"[culture clash, future, space war, space colon...","In the 22nd century, a paraplegic Marine is di...",2009,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,"[Adventure, Fantasy, Action]",285,"[ocean, drug abuse, exotic island, east india ...","Captain Barbossa, long believed to be dead, ha...",2007,Pirates of the Caribbean: At World's End,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,"[Action, Adventure, Crime]",206647,"[spy, based on novel, secret agent, sequel, mi...",A cryptic message from Bond’s past sends him o...,2015,Spectre,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,"[Action, Crime, Drama, Thriller]",49026,"[dc comics, crime fighter, terrorist, secret i...",Following the death of District Attorney Harve...,2012,The Dark Knight Rises,"[{""cast_id"": 2, ""character"": ""Bruce Wayne / Ba...","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,"[Action, Adventure, Science Fiction]",49529,"[based on novel, mars, medallion, space travel...","John Carter is a war-weary, former military ca...",2012,John Carter,"[{""cast_id"": 5, ""character"": ""John Carter"", ""c...","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."


## Processing Cast Column

The **cast** column contains information about the actors who appeared in the movie.

The data is stored in a **JSON-like format**, where each entry contains details such as:
- cast_id
- character
- name
- gender

For building the recommendation system, we only need the **names of the actors**.

To keep the dataset simple and focused, we usually extract **only the top 3 main actors** from the cast list.

In [31]:
data["cast"][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [32]:
import ast

def extract_cast(text):
    actors = []
    counter = 0
    
    for i in ast.literal_eval(text):
        if counter < 3:
            actors.append(i['name'])
            counter += 1
        else:
            break
            
    return actors

data['cast'] = data['cast'].apply(extract_cast)

In [33]:
data["cast"][0]

['Sam Worthington', 'Zoe Saldana', 'Sigourney Weaver']

## Processing Crew Column

The **crew** column contains information about the crew members involved in the movie, such as:

- Director
- Producer
- Writer
- Cinematographer
- Editor

For building a recommendation system, the **director is the most important crew member** because the director strongly influences the style and theme of the movie.

Therefore, we extract only the **director name** from the crew column.

In [34]:
data["crew"][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [35]:
import ast

def extract_director(text):
    directors = []
    
    for i in ast.literal_eval(text):
        if i['job'] == 'Director':
            directors.append(i['name'])
            break
            
    return directors

data['crew'] = data['crew'].apply(extract_director)

In [36]:
data.head()

,genres,id,keywords,overview,release_date,title,cast,crew
0,"[Action, Adventure, Fantasy, Science Fiction]",19995,"[culture clash, future, space war, space colon...","In the 22nd century, a paraplegic Marine is di...",2009,Avatar,"[Sam Worthington, Zoe Saldana, Sigourney Weaver]",[James Cameron]
1,"[Adventure, Fantasy, Action]",285,"[ocean, drug abuse, exotic island, east india ...","Captain Barbossa, long believed to be dead, ha...",2007,Pirates of the Caribbean: At World's End,"[Johnny Depp, Orlando Bloom, Keira Knightley]",[Gore Verbinski]
2,"[Action, Adventure, Crime]",206647,"[spy, based on novel, secret agent, sequel, mi...",A cryptic message from Bond’s past sends him o...,2015,Spectre,"[Daniel Craig, Christoph Waltz, Léa Seydoux]",[Sam Mendes]
3,"[Action, Crime, Drama, Thriller]",49026,"[dc comics, crime fighter, terrorist, secret i...",Following the death of District Attorney Harve...,2012,The Dark Knight Rises,"[Christian Bale, Michael Caine, Gary Oldman]",[Christopher Nolan]
4,"[Action, Adventure, Science Fiction]",49529,"[based on novel, mars, medallion, space travel...","John Carter is a war-weary, former military ca...",2012,John Carter,"[Taylor Kitsch, Lynn Collins, Samantha Morton]",[Andrew Stanton]


## Processing Overview Column

The **overview** column contains a short description or summary of the movie's storyline.

This column is very important because it provides information about the **plot, theme, and context of the movie**.

To use this information in the recommendation system, we convert the overview text into a **list of words** so it can later be combined with other features such as genres, keywords, cast, and director.

This process helps in creating the **tags column**, which represents the overall content of the movie.

In [37]:
data['overview'] = data['overview'].apply(lambda x: x.split())

In [38]:
data.head()

,genres,id,keywords,overview,release_date,title,cast,crew
0,"[Action, Adventure, Fantasy, Science Fiction]",19995,"[culture clash, future, space war, space colon...","[In, the, 22nd, century,, a, paraplegic, Marin...",2009,Avatar,"[Sam Worthington, Zoe Saldana, Sigourney Weaver]",[James Cameron]
1,"[Adventure, Fantasy, Action]",285,"[ocean, drug abuse, exotic island, east india ...","[Captain, Barbossa,, long, believed, to, be, d...",2007,Pirates of the Caribbean: At World's End,"[Johnny Depp, Orlando Bloom, Keira Knightley]",[Gore Verbinski]
2,"[Action, Adventure, Crime]",206647,"[spy, based on novel, secret agent, sequel, mi...","[A, cryptic, message, from, Bond’s, past, send...",2015,Spectre,"[Daniel Craig, Christoph Waltz, Léa Seydoux]",[Sam Mendes]
3,"[Action, Crime, Drama, Thriller]",49026,"[dc comics, crime fighter, terrorist, secret i...","[Following, the, death, of, District, Attorney...",2012,The Dark Knight Rises,"[Christian Bale, Michael Caine, Gary Oldman]",[Christopher Nolan]
4,"[Action, Adventure, Science Fiction]",49529,"[based on novel, mars, medallion, space travel...","[John, Carter, is, a, war-weary,, former, mili...",2012,John Carter,"[Taylor Kitsch, Lynn Collins, Samantha Morton]",[Andrew Stanton]


## Removing Spaces Between Names

In the Movie Recommendation System, we remove spaces between names so that multi-word names are treated as a **single token** during text vectorization.

If spaces are not removed, the model may treat **first name and last name as two different words**, which can affect the accuracy of similarity calculations.

### Example

Before removing spaces:

Sam Worthington → ['Sam', 'Worthington']

After removing spaces:

SamWorthington

Now the model treats it as **one unique feature**, which improves the quality of the recommendation system.

### Columns Where We Remove Spaces

We usually remove spaces from the following columns:

- genres
- keywords
- cast
- crew

In [39]:
# Remove spaces between words in selected columns

data['genres'] = data['genres'].apply(lambda x: [i.replace(" ", "") for i in x])
data['keywords'] = data['keywords'].apply(lambda x: [i.replace(" ", "") for i in x])
data['cast'] = data['cast'].apply(lambda x: [i.replace(" ", "") for i in x])
data['crew'] = data['crew'].apply(lambda x: [i.replace(" ", "") for i in x])

In [40]:
data.head()

,genres,id,keywords,overview,release_date,title,cast,crew
0,"[Action, Adventure, Fantasy, ScienceFiction]",19995,"[cultureclash, future, spacewar, spacecolony, ...","[In, the, 22nd, century,, a, paraplegic, Marin...",2009,Avatar,"[SamWorthington, ZoeSaldana, SigourneyWeaver]",[JamesCameron]
1,"[Adventure, Fantasy, Action]",285,"[ocean, drugabuse, exoticisland, eastindiatrad...","[Captain, Barbossa,, long, believed, to, be, d...",2007,Pirates of the Caribbean: At World's End,"[JohnnyDepp, OrlandoBloom, KeiraKnightley]",[GoreVerbinski]
2,"[Action, Adventure, Crime]",206647,"[spy, basedonnovel, secretagent, sequel, mi6, ...","[A, cryptic, message, from, Bond’s, past, send...",2015,Spectre,"[DanielCraig, ChristophWaltz, LéaSeydoux]",[SamMendes]
3,"[Action, Crime, Drama, Thriller]",49026,"[dccomics, crimefighter, terrorist, secretiden...","[Following, the, death, of, District, Attorney...",2012,The Dark Knight Rises,"[ChristianBale, MichaelCaine, GaryOldman]",[ChristopherNolan]
4,"[Action, Adventure, ScienceFiction]",49529,"[basedonnovel, mars, medallion, spacetravel, p...","[John, Carter, is, a, war-weary,, former, mili...",2012,John Carter,"[TaylorKitsch, LynnCollins, SamanthaMorton]",[AndrewStanton]


## Creating the Tags Column

The **tags column** is created by combining important text features of the movie into a single column.

These features include:

- overview
- genres
- keywords
- cast
- crew (director)

Combining these features helps represent the **overall content of the movie**.

The tags column will later be used for **text vectorization and similarity calculation** to recommend similar movies.

In [41]:
# Create tags column by combining important features
data['tags'] = data['overview'] + data['genres'] + data['keywords'] + data['cast'] + data['crew']

In [42]:
data[['title','tags']].head()

,title,tags
0,Avatar,"[In, the, 22nd, century,, a, paraplegic, Marin..."
1,Pirates of the Caribbean: At World's End,"[Captain, Barbossa,, long, believed, to, be, d..."
2,Spectre,"[A, cryptic, message, from, Bond’s, past, send..."
3,The Dark Knight Rises,"[Following, the, death, of, District, Attorney..."
4,John Carter,"[John, Carter, is, a, war-weary,, former, mili..."


In [43]:
new_df = data[['id','title','tags']]

In [44]:
new_df.head()

,id,title,tags
0,19995,Avatar,"[In, the, 22nd, century,, a, paraplegic, Marin..."
1,285,Pirates of the Caribbean: At World's End,"[Captain, Barbossa,, long, believed, to, be, d..."
2,206647,Spectre,"[A, cryptic, message, from, Bond’s, past, send..."
3,49026,The Dark Knight Rises,"[Following, the, death, of, District, Attorney..."
4,49529,John Carter,"[John, Carter, is, a, war-weary,, former, mili..."


# Next important steps in the recommender system:

- Convert tags list → string

- Convert text to lowercase

- Apply CountVectorizer (Bag of Words)

- Calculate Cosine Similarity

- Create the recommendation function

## Converting Tags List to String

After creating the **tags column**, the values are stored as a list of words.

Example:

['action', 'adventure', 'samworthington', 'jamescameron']

However, text vectorization techniques such as **Bag of Words (CountVectorizer)** require the input to be in **string format**.

Therefore, we convert the list into a single string.

---

## Converting Text to Lowercase

After converting the list into a string, we convert all text to **lowercase**.

This step ensures that words like **Action, action, ACTION** are treated as the same word during vectorization.

In [45]:
# Convert tags list to string
new_df['tags'] = new_df['tags'].apply(lambda x: " ".join(x))

# Convert text to lowercase
new_df['tags'] = new_df['tags'].apply(lambda x: x.lower())
new_df['title'] = new_df['title'].apply(lambda x: x.lower())

C:\Users\AS00\AppData\Local\Temp\ipykernel_16496\2495962492.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x: " ".join(x))
C:\Users\AS00\AppData\Local\Temp\ipykernel_16496\2495962492.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x: x.lower())
C:\Users\AS00\AppData\Local\Temp\ipykernel_16496\2495962492.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using 

In [46]:
new_df.head()

,id,title,tags
0,19995,avatar,"in the 22nd century, a paraplegic marine is di..."
1,285,pirates of the caribbean: at world's end,"captain barbossa, long believed to be dead, ha..."
2,206647,spectre,a cryptic message from bond’s past sends him o...
3,49026,the dark knight rises,following the death of district attorney harve...
4,49529,john carter,"john carter is a war-weary, former military ca..."


## Next Steps :

After creating the **tags column**, we will perform the following steps to build the recommendation system.

### 1️⃣ Remove Stop Words

Stop words are common words that do not add meaningful information to the text.

Examples:
- the
- is
- and
- in
- of

Removing these words helps reduce noise in the dataset and improves model performance.

Example:

Before:
the hero saves the world

After removing stop words:
hero saves world

---

### 2️⃣ Stemming

Stemming reduces words to their **root/base form**.

Example:

| Original Word | Stemmed Word |
|---------------|-------------|
| loving | love |
| loved | love |
| loves | love |
| playing | play |

This helps treat similar words as the **same feature** in the model.

---

### 3️⃣ Bag of Words (Text Vectorization)

After preprocessing the text, we convert the text into **numerical vectors** using **Bag of Words (BoW)**.

Bag of Words creates a vocabulary of all unique words and represents each movie as a **vector of word frequencies**.

We use **CountVectorizer** from sklearn for this step.

---

### 4️⃣ Cosine Similarity

Cosine similarity measures how similar two movie vectors are.

It calculates the **angle between two vectors**.

Value Range:

- **1 → Highly Similar**
- **0 → Not Similar**

This creates a **similarity matrix** where each movie is compared with every other movie.

---

### 5️⃣ Movie Recommendation Function

Finally, we build a function that recommends movies.

Steps:
1. Take a **movie name** as input
2. Find the **index of the movie**
3. Retrieve similarity scores
4. Sort movies based on similarity
5. Return the **top 5 most similar movies**

This function forms the **core logic of the recommendation system**.

In [47]:
# Remove Stop Words

from sklearn.feature_extraction.text import CountVectorizer

# Stop words will automatically be removed when using stop_words='english' in CountVectorizer later.

- Does NOT change the tags column

- Removes stop words only while creating vectors

- Stop words are ignored during feature creation

In [ ]:
cv = CountVectorizer(max_features=5000, stop_words='english')






(4805, 5000)
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [67]:
new_df.head()

,id,title,tags
0,19995,avatar,"in the 22nd century, a parapleg marin is dispa..."
1,285,pirates of the caribbean: at world's end,"captain barbossa, long believ to be dead, ha c..."
2,206647,spectre,a cryptic messag from bond’ past send him on a...
3,49026,the dark knight rises,follow the death of district attorney harvey d...
4,49529,john carter,"john carter is a war-weary, former militari ca..."


In [68]:
# Stemming 

import nltk
from nltk.stem.porter import PorterStemmer

ps = PorterStemmer()

In [69]:
# Create a stemming function 
#  
def stem(text):
    y = []
    
    for i in text.split():
        y.append(ps.stem(i))
        
    return " ".join(y)

In [70]:
# Apply stemming to tags column

new_df['tags'] = new_df['tags'].apply(stem)

C:\Users\AS00\AppData\Local\Temp\ipykernel_16496\2922227616.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(stem)


In [72]:
new_df["tags"][0]

'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'

In [73]:
new_df.head()

,id,title,tags
0,19995,avatar,"in the 22nd century, a parapleg marin is dispa..."
1,285,pirates of the caribbean: at world's end,"captain barbossa, long believ to be dead, ha c..."
2,206647,spectre,a cryptic messag from bond’ past send him on a...
3,49026,the dark knight rises,follow the death of district attorney harvey d...
4,49529,john carter,"john carter is a war-weary, former militari ca..."




###  Text Vectorization (Bag of Words / CountVectorizer)


After completing preprocessing steps such as **stop word removal** and **stemming**, the next step is to convert text data into numerical form so that machine learning algorithms can understand it.

This process is called **Text Vectorization**.

One of the most commonly used techniques is **Bag of Words (BoW)**.

### What is Bag of Words?

Bag of Words converts text into a **vector of word frequencies**.

Steps:
1. Create a **vocabulary** of the most frequent words in the dataset.
2. Count how many times each word appears in each document.
3. Represent each document as a **numerical vector**.

### Example

Sentence 1:  
I love machine learning

Sentence 2:  
I love NLP

Vocabulary:

I | love | machine | learning | NLP

Vector Representation:

Sentence 1 → [1,1,1,1,0]  
Sentence 2 → [1,1,0,0,1]

In this project, we use **CountVectorizer from sklearn** to implement Bag of Words.

## Why We Use CountVectorizer to Implement Bag of Words

In Natural Language Processing (NLP), machine learning models cannot understand raw text directly.  
Therefore, we need to convert text into **numerical vectors**. One common technique for this is **Bag of Words (BoW)**.

### Why CountVectorizer?

`CountVectorizer` from **sklearn** is used to implement Bag of Words because it automatically performs several important tasks:

### 1️⃣ Vocabulary Creation
It scans the entire dataset and creates a **vocabulary of unique words**.

Example:

Text Data:
- "I love movies"
- "I love NLP"

Vocabulary created:
I | love | movies | NLP

---

### 2️⃣ Word Frequency Counting
It counts how many times each word appears in each document.

Example:

| Sentence | I | love | movies | NLP |
|--------|---|---|---|---|
| I love movies | 1 | 1 | 1 | 0 |
| I love NLP | 1 | 1 | 0 | 1 |

---

### 3️⃣ Converts Text into Numerical Matrix
It converts the text data into a **matrix of numbers** that machine learning algorithms can process.

Output shape example:

(4806 movies, 5000 words)

Each movie is represented as a **vector of word frequencies**.

---

### 4️⃣ Built-in Text Processing
`CountVectorizer` can also perform useful preprocessing automatically:

- Convert text to **lowercase**
- **Remove stop words**
- **Limit vocabulary size** using `max_features`
- Tokenize words automatically

Example:

```python
cv = CountVectorizer(max_features=5000, stop_words='english')

Computers cannot understand text directly, so we convert the **tags text data into numerical vectors**.

We use **Bag of Words (BoW)** with **CountVectorizer** to transform the text into a matrix of token counts.

This step creates a **feature matrix** where each movie is represented as a numerical vector.

---

###  Calculate Cosine Similarity

Once the movies are converted into vectors, we calculate the **similarity between movies**.

We use **Cosine Similarity** to measure how similar two movie vectors are.

- Value close to **1 → Highly Similar**
- Value close to **0 → Not Similar**

This creates a **similarity matrix** comparing every movie with every other movie.

---

###  Build the Recommendation Function

Finally, we create a **recommendation function** that:

1. Takes a **movie title as input**
2. Finds the **similarity scores** for that movie
3. Sorts movies based on similarity
4. Returns the **top 5 most similar movies**

This function forms the **core of the movie recommendation system**.

In [74]:
## Bag of words 

vectors = cv.fit_transform(new_df['tags']).toarray()

vectors = cv.fit_transform(new_df['tags'])



In [75]:
vectors.shape

(4805, 5000)

In [ ]:
# see vocabulary words
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zone', 'zoo', 'zooeydeschanel'],
      shape=(5000,), dtype=object)

In [87]:
# chech vector
vectors = vectors.toarray()
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], shape=(4805, 5000))

In [88]:
cv.get_feature_names_out()[:200]

array(['000', '007', '10', '100', '11', '12', '13', '14', '15', '16',
       '17', '17th', '18', '18th', '18thcenturi', '19', '1910', '1920',
       '1930', '1940', '1944', '1950', '1950s', '1960', '1960s', '1970',
       '1970s', '1974', '1976', '1980', '1985', '1990', '1999', '19th',
       '19thcenturi', '20', '200', '2003', '2009', '20th', '21st', '23',
       '24', '25', '30', '300', '3d', '40', '50', '500', '60', '70', '80',
       'aaron', 'aaroneckhart', 'abandon', 'abduct', 'abigailbreslin',
       'abil', 'abl', 'aboard', 'abov', 'abu', 'academ', 'academi',
       'accept', 'access', 'accid', 'accident', 'acclaim', 'accompani',
       'accomplish', 'account', 'accu', 'ace', 'achiev', 'acquaint',
       'act', 'action', 'actionhero', 'activ', 'activist', 'activities',
       'actor', 'actress', 'actual', 'ad', 'adam', 'adamsandl',
       'adamshankman', 'adapt', 'add', 'addict', 'adjust', 'admir',
       'admit', 'adolesc', 'adopt', 'ador', 'adrienbrodi', 'adult',
       'adul

## Cosine Similarity

After converting text data into numerical vectors using **Bag of Words**, we need a method to measure how similar two movies are.

For this purpose, we use **Cosine Similarity**.

### What is Cosine Similarity?

Cosine Similarity is a metric used to measure the **similarity between two vectors** by calculating the **cosine of the angle between them**.

Instead of comparing the magnitude of the vectors, it compares their **direction**.

### Cosine Similarity Formula

Cosine Similarity = (A · B) / (||A|| × ||B||)

Where:
- **A · B** → Dot product of vectors A and B
- **||A||** → Magnitude of vector A
- **||B||** → Magnitude of vector B

### Cosine Similarity Range

| Value | Meaning |
|------|------|
| 1 | Very similar |
| 0 | Not similar |
| -1 | Completely opposite |

In a **movie recommendation system**, we compute cosine similarity between movie vectors to find movies that are most similar to each other.

Movies with **higher similarity scores** are recommended to the user.

In [89]:
from sklearn.metrics.pairwise import cosine_similarity

similarity = cosine_similarity(vectors)
similarity.shape


(4805, 4805)

In [90]:
similarity

array([[1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
        0.        ],
       [0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
        0.02615329],
       [0.0860309 , 0.06063391, 1.        , ..., 0.02451452, 0.        ,
        0.        ],
       ...,
       [0.04499213, 0.02378257, 0.02451452, ..., 1.        , 0.03962144,
        0.04229549],
       [0.        , 0.        , 0.        , ..., 0.03962144, 1.        ,
        0.08714204],
       [0.        , 0.02615329, 0.        , ..., 0.04229549, 0.08714204,
        1.        ]], shape=(4805, 4805))

## Movie Recommendation Function

After computing the **Cosine Similarity matrix**, the next step is to build a function that recommends movies based on similarity scores.

### How the Recommendation Function Works

1. Take a **movie title** as input.
2. Find the **index of the movie** in the dataset.
3. Retrieve the **similarity scores** for that movie.
4. Sort the movies based on **highest similarity score**.
5. Return the **top 5 most similar movies**.

Movies with the **highest cosine similarity scores** are considered the most similar and are recommended to the user.

In [91]:
def recommend(movie):
    
    # find the index of the movie
    movie_index = new_df[new_df['title'] == movie].index[0]
    
    # get similarity scores for that movie
    distances = similarity[movie_index]
    
    # sort movies based on similarity score
    movies_list = sorted(list(enumerate(distances)), reverse=True, key=lambda x: x[1])[1:6]
    
    # print recommended movie titles
    for i in movies_list:
        print(new_df.iloc[i[0]].title)

In [92]:
recommend("avatar")

aliens vs predator: requiem
aliens
falcon rising
independence day
titan a.e.


## Saving the Model Using Pickle

After building the recommendation system, we need to save the processed data and similarity matrix so that we can use them later without retraining the model.

For this purpose, we use **Pickle**, a Python module that allows us to serialize and store Python objects.

We will save:
1. The movie dataset
2. The similarity matrix

These files will later be used in the **Streamlit application** to generate recommendations.

In [93]:
import pickle

pickle.dump(new_df, open('movies.pkl','wb'))
pickle.dump(similarity, open('similarity.pkl','wb'))